File: SwiftBoy/cpp/GameBoy.cpp
//
//  GameBoy.cpp
//  SwiftBoy
//
//  Created by Andrew Gilliland on 12/21/25.
//

#include "GameBoy.hpp"
#include <iostream>
#include <cstring>

GameBoy::GameBoy() 
    : audioSampleCount(0), buttonState(0), romLoaded(false) {
    std::cout << "[GameBoy] Constructor called" << std::endl;
    
    // Initialize frame buffer to a test pattern (alternating red/green)
    for (int i = 0; i < FRAME_BUFFER_SIZE; i += 4) {
        frameBuffer[i + 0] = (i / 4) % 2 == 0 ? 255 : 0;  // R
        frameBuffer[i + 1] = (i / 4) % 2 == 0 ? 0 : 255;  // G
        frameBuffer[i + 2] = 0;                            // B
        frameBuffer[i + 3] = 255;                          // A
    }
    
    // Initialize audio buffer to silence
    std::memset(audioBuffer, 0, sizeof(audioBuffer));
}

GameBoy::~GameBoy() {
    std::cout << "[GameBoy] Destructor called" << std::endl;
}

bool GameBoy::loadROM(const uint8_t* data, std::size_t size) {
    std::cout << "[GameBoy] Loading ROM: " << size << " bytes" << std::endl;
    
    if (data == nullptr || size == 0) {
        std::cerr << "[GameBoy] ERROR: Invalid ROM data" << std::endl;
        return false;
    }
    
    // TODO: Actually load ROM data into memory
    romLoaded = true;
    std::cout << "[GameBoy] ROM loaded successfully" << std::endl;
    return true;
}

void GameBoy::runFrame() {
    if (!romLoaded) {
        std::cout << "[GameBoy] WARNING: runFrame() called but no ROM loaded" << std::endl;
        return;
    }
    
    std::cout << "[GameBoy] Running frame (~70,224 CPU cycles)" << std::endl;
    
    // TODO: Execute CPU cycles
    // TODO: Update PPU (render scanlines)
    // TODO: Update APU (generate audio samples)
    // TODO: Update timers
    // TODO: Handle interrupts
    
    // For now, just cycle through colors in the frame buffer
    static uint8_t colorShift = 0;
    colorShift++;
    for (int i = 0; i < FRAME_BUFFER_SIZE; i += 4) {
        frameBuffer[i + 0] = (i / 4 + colorShift) % 255;  // R
        frameBuffer[i + 1] = (i / 4 * 2 + colorShift) % 255;  // G
        frameBuffer[i + 2] = (i / 4 * 3 + colorShift) % 255;  // B
        frameBuffer[i + 3] = 255;  // A
    }
}

const uint8_t* GameBoy::getFrameBuffer() {
    std::cout << "[GameBoy] getFrameBuffer() called" << std::endl;
    return frameBuffer;
}

const int16_t* GameBoy::getAudioBuffer(std::size_t* outSize) {
    std::cout << "[GameBoy] getAudioBuffer() called" << std::endl;
    
    if (outSize != nullptr) {
        *outSize = audioSampleCount;
    }
    
    // TODO: Generate actual audio samples
    // For now, return silence
    audioSampleCount = 0;
    return audioBuffer;
}

void GameBoy::setButtonState(uint8_t buttons) {
    std::cout << "[GameBoy] Button state: 0x" << std::hex << (int)buttons << std::dec << std::endl;
    buttonState = buttons;
    
    // TODO: Update joypad registers
    // TODO: Trigger joypad interrupt if needed
}

void GameBoy::reset() {
    std::cout << "[GameBoy] Reset called" << std::endl;
    
    buttonState = 0;
    audioSampleCount = 0;
    
    // TODO: Reset CPU state
    // TODO: Reset memory
    // TODO: Reset PPU
    // TODO: Reset APU
    // TODO: Reset timers
    
    std::cout << "[GameBoy] Reset complete" << std::endl;
}

File: SwiftBoy/GameBoy/GameBoy.swift
//
//  GameBoy.swift
//  SwiftBoy
//
//  Created by Andrew Gilliland on 12/25/25.
//

import Foundation

class GameBoy {
    // Hardware components
    private let memory: Memory
    // TODO: Add CPU, PPU, APU, Timer, Joypad
    
    // Display
    private static let screenWidth = 160
    private static let screenHeight = 144
    private static let bytesPerPixel = 4  // RGBA
    private static let frameBufferSize = screenWidth * screenHeight * bytesPerPixel
    
    private var frameBuffer: [UInt8]
    
    // Audio
    private static let audioBufferSize = 2048  // Stereo samples
    private var audioBuffer: [Int16]
    private var audioSampleCount: Int = 0
    
    // Input
    private var buttonState: UInt8 = 0
    
    // State
    private var romLoaded: Bool = false
    
    init() {
        print("[GameBoy] Initializing...")
        
        // Initialize memory system
        self.memory = Memory()
        
        // Initialize frame buffer with test pattern (alternating red/green)
        self.frameBuffer = [UInt8](repeating: 0, count: Self.frameBufferSize)
        for i in stride(from: 0, to: Self.frameBufferSize, by: 4) {
            let pixelIndex = i / 4
            frameBuffer[i + 0] = (pixelIndex % 2 == 0) ? 255 : 0  // R
            frameBuffer[i + 1] = (pixelIndex % 2 == 0) ? 0 : 255  // G
            frameBuffer[i + 2] = 0                                 // B
            frameBuffer[i + 3] = 255                               // A
        }
        
        // Initialize audio buffer with silence
        self.audioBuffer = [Int16](repeating: 0, count: Self.audioBufferSize)
        
        print("[GameBoy] Initialized")
    }
    
    deinit {
        print("[GameBoy] Deallocating")
    }
    
    // MARK: - ROM Loading
    
    func loadROM(_ data: Data) -> Bool {
        print("[GameBoy] Loading ROM: \(data.count) bytes")
        
        guard !data.isEmpty else {
            print("[GameBoy] ERROR: Empty ROM data")
            return false
        }
        
        // Load ROM into memory system
        let success = memory.loadROM(data)
        romLoaded = success
        
        if success {
            print("[GameBoy] ROM loaded successfully")
            
            // Verify ROM data
            let entryPoint = memory.read(address: 0x0100)
            print("[GameBoy] ROM entry point (0x0100): 0x\(String(entryPoint, radix: 16))")
            
            // TODO: Reset CPU after ROM load
            print("[GameBoy] Ready to run")
        } else {
            print("[GameBoy] ERROR: Failed to load ROM")
        }
        
        return success
    }
    
    // MARK: - Emulation
    
    func runFrame() {
        guard romLoaded else {
            print("[GameBoy] WARNING: runFrame() called but no ROM loaded")
            return
        }
        
        // Execute ~70,224 CPU cycles (one frame at 4.194 MHz, 60 Hz)
        // TODO: Implement CPU execution
        // TODO: Update PPU (render scanlines)
        // TODO: Update APU (generate audio samples)
        // TODO: Update timers
        // TODO: Handle interrupts
    }
    
    func reset() {
        print("[GameBoy] Resetting...")
        
        buttonState = 0
        audioSampleCount = 0
        
        // TODO: Reset CPU
        // TODO: Reset PPU
        // TODO: Reset APU
        // TODO: Reset timers
        
        print("[GameBoy] Reset complete")
    }
    
    // MARK: - Output
    
    func getFrameBuffer() -> UnsafePointer<UInt8> {
        return frameBuffer.withUnsafeBufferPointer { $0.baseAddress! }
    }
    
    func getAudioBuffer() -> (UnsafePointer<Int16>, Int) {
        // TODO: Generate actual audio samples
        audioSampleCount = 0
        return audioBuffer.withUnsafeBufferPointer { ($0.baseAddress!, audioSampleCount) }
    }
    
    // MARK: - Input
    
    func setButtonState(_ buttons: UInt8) {
        print("[GameBoy] Button state: 0x\(String(buttons, radix: 16))")
        buttonState = buttons
        
        // TODO: Update joypad registers
        // TODO: Trigger joypad interrupt if needed
    }
}
